tests/31sync/09archived.pl

```diff
@@ -0,0 +1,193 @@
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id, $next );
+
+     matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         matrix_create_room( $user );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
+
+         matrix_leave_room( $user, $room_id );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id, since => $next );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Newly left rooms appear in the archived section of gapped sync",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user, $filter_id, $room_id_1, $room_id_2, $next );
+
+      matrix_register_user_with_filter( $http, {} )->then( sub {
+         ( $user, $filter_id ) = @_;
+
+         Future->needs_all(
+            matrix_create_room( $user )->on_done( sub { ( $room_id_1 ) = @_; } ),
+            matrix_create_room( $user )->on_done( sub { ( $room_id_2 ) = @_; } ),
+         );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
+
+         matrix_leave_room( $user, $room_id_1 );
+      })->then( sub {
+         Future->needs_all( map {
+            matrix_send_room_message( $user, $room_id_2,
+               content => { "filler" => $_ },
+               type    => "a.made.up.filler.type",
+            )
+         } 0 .. 20 );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id, since => $next );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id_1};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Archived rooms only contain history from before the user left",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user_a, $filter_id_a, $user_b, $filter_id_b, $room_id, $next_b );
+
+      my $filter = {
+         room => {
+            timeline => { types => [ "m.room.message" ] },
+            state => { types => [ "a.madeup.test.state" ] },
+         },
+      };
+
+      Future->needs_all(
+         matrix_register_user_with_filter( $http, $filter ),
+         matrix_register_user_with_filter( $http, $filter ),
+      )->then( sub {
+         ( $user_a, $filter_id_a, $user_b, $filter_id_b ) = @_;
+
+         matrix_create_room( $user_a );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_invite_user_to_room( $user_a, $user_b, $room_id );
```

Comment:
Eh. I wasn't sure if that was a documented, intentional behaviour of matrix_create_room.